api/src/main/java/org/geysermc/geyser/api/pack/PackCodec.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2019-2023 GeyserMC. http://geysermc.org
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.pack;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.geysermc.geyser.api.GeyserApi;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Path;
+
+/**
+ * Represents a pack codec that can be used
+ * to provide resource packs to clients.
+ */
+public abstract class PackCodec {
```

Comment:
Could just be an interface? Same with PathPackCodec

Reply:
Has a protected abstract method so that’s not possible unfortunately.

Reply:
Just so that people have to call it on ResourcePack? I'm not saying the method shouldn't be in ResourcePack, but all it does is directly call `create()`

Reply:
We could expose it that way, but it'd mean there's duplicate API. We could also go the route of not having `create` in ResourcePack, but it's definitely far nicer than doing `PackCodec.path(Paths.get(...)).create()`

Reply:
Fair. And I guess keeping it abstract would allow use to add state/behaviour in the future that we don't want to expose publicly